### ISSUE TITLE
Reconsider defensive-style implementation

### ISSUE THREAD

**Issue by HappySeaFox** [HappySeaFox/sail#152]
E.g. when every function checks the input parameters.

**Comment by JoniSt** [comment]
One thing that comes to my mind about defensive programming, especially argument checking, is this: Is the error checking code checking for a bug in the programmer's code? (Either SAIL's code or the code that's using SAIL.) If so, then the error check should most likely be an assertion that immediately terminates the program if it fails. Otherwise it'll make bugs harder to fix, rather than easier.

My reasoning behind that is this: If a programmer doesn't expect a particular function to fail with the input they give to it, they won't do error checking. Suppose now that the programmer wrote a bug and passed incorrect arguments - a "defensive style" function will realize that and abort the operation, possibly returning an error code. That means an immediate crash has been avoided - but the code that called the function will be thoroughly confused because one of its vital operations just didn't take place. That, in turn, is going to corrupt the program's state and possibly make it crash later (or worse, silently corrupt some data).

In my opinion, if a function has the opportunity to (easily) validate some input data or check the integrity of a data structure, it should do so. However, if the validation fails, it _shouldn't return an error code_ but instead _crash as early as possible_. No error handling, no cleanup, nothing - just a plain "if this assertion doesn't hold, the program is already broken and going to crash anyway. Crash it now to make finding the bug easier (= call ```abort()```)." The programmer can then easily place a breakpoint on the ```abort``` function and see the exact point in time where the program messed up instead of having to trace backwards through the code.

If the programmer realizes that an invalid argument / invalid state like this can occur naturally in their program and isn't a bug, they can still validate the arguments manually _before_ they call the function. That way, it's made explicit in their code.

Assertions also have the advantage that they can be disabled if maximum performance is desired. This could either be a compile-time configuration or a runtime setting. Given that SAIL is a library that's intended to be shipped in binary form with headers, a runtime setting probably makes the most sense. That could be as simple as a ```sail_disable_expensive_assertions()``` function that the user can call if they want more performance. Cheap checks like bounds/null checks probably shouldn't be disabled either way: CPUs are so good at speculative execution that such checks are literally free. The check is just going to be done on an otherwise-idle resource (i.e. ALU) and the branch will always be predicted correctly (because it always goes the same way). Additionally, checking whether assertions are enabled (assuming the runtime variant is used) is just as expensive as doing the actual assertion: Both are a comparison and a conditional jump.

Of course, if a particular error condition can be caused by user input (i.e. a malformed image), that should still be detected and reported via error codes since it doesn't imply that there's a bug in the program. The program is expected to handle this kind of error.

TL;DR: Try to detect bugs, but don't try to "correct" them silently. Instead, crash the program immediately with a descriptive message so that the programmer can fix them.

**Comment by HappySeaFox** [comment]
Thanks 👍 Absolutely great investigation. My 2 cents. We can avoid using `abort` and naturally crash. Segfaults are easily detectible. Even more, nowadays it's easy to detect any memory-related issue with ASAN if the user wants to. So we don't pollute the code with `abort`.

**Comment by HappySeaFox** [comment]
I recently committed a couple of changes to defensive functions. See https://github.com/search?q=defensive+user%3AHappySeaFox&type=Commits .

If you have more functions in mind that can be made non-defensive, please let me know.
